docs: added gRPC custom name resolver proposal (#1414)
open-feature/flagd: docs/reference/specifications/proposal/rfc-grpc-custom-name-resolver.md

    flagd-provider-->>application: Get the feature flag from in-memory cache <br/> run the evaluation logic and return final state
    loop
    flagd-provider->>flagd-provider: in-memory cache
    end
    flagd-provider->>proxy-sidecar-agent: gRPC stream connection
    proxy-sidecar-agent-->>flagd-provider: 
    Note right of flagd-provider: Instead host:port target string <br> "envoy://localhost:9211/flagd-sync.service" <br> will be used
    proxy-sidecar-agent->>flagd-sync.service: Apply required policy and route traffic <br> to backend nodes
    flagd-sync.service-->>proxy-sidecar-agent: 
    Note right of proxy-sidecar-agent: Policy and route rules are applied based <br> on `authority` header used by the <br> gRPC client
```

#### Drawbacks

* One of the big drawback was limited support of the language only `java` and `golang`
* Will introduce inconsistent user experience
* Will open the door for different use cases although this can be fixed by
providing sdks similar to [custom connector](https://github.com/open-feature/java-sdk-contrib/tree/main/providers/flagd#custom-connector)
* ...

## Alternatives

### Option-1

> Allow users to override default ``authority`` header as shown above in `grpcurl`, the override option was

Reviewer: ```suggestion
Allow users to override default `authority` header as shown above in `grpcurl`, the override option was
```

# gRPC Custom Name Resolver Proposal (DRAFT)

## Details

|                        |                                    |
|------------------------|------------------------------------|
| **Feature Name**       | gRPC custom name resolver          |
| **Type**               | enhancement                        |
| **Related components** | gRPC source resolution             |

## Summary

> gRPC by default support dns resolution which is currently being used e.g. "localhost:8013" in both

Reviewer 2: ```suggestion
gRPC by default supports DNS resolution which is currently being used e.g. "localhost:8013" in both
```

[core](https://github.com/open-feature/flagd/blob/main/core/pkg/sync/grpc/grpc_sync.go#L72-L74) and
providers e.g. [java](https://github.com/open-feature/java-sdk-contrib/blob/main/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/common/ChannelBuilder.java#L53-L55).

> This covers most deployments but with increase adoption of microservice architecture and service discovery
> application with S2S policy e.g. istio, envoy, consul etc. In such deployments the current gRPC connection was
> breaking because the proxies need the destination domain i.e. service name in order to route the traffic.

Reviewer 2: ```suggestion
This covers most deployments, but with increased adoption of microservice-architecture, service discovery,
policy-enabled service meshes (e.g. istio, envoy, consul, etc) it's necessary to support custom routing and name resolution.
```

> For such cases the gRPC core libs support few alternative resolver* also expose required interface to build custom one
> you can find more details in the document referenced below
> 

Reviewer 2: ```suggestion
For such cases the gRPC core libraries support few alternative resolver* also expose the required interfaces to build custom implementations:

```

* [Java Client](https://grpc.github.io/grpc-java/javadoc/io/grpc/ManagedChannelBuilder.html#forTarget(java.lang.String))
* [Golang](https://pkg.go.dev/google.golang.org/grpc#NewClient)

A sample workflow of a deployment with proxy sidecar using `in-process` mode (for `rpc` the process is same)

```mermaid
sequenceDiagram
    participant flagd-sync.service
    participant proxy-sidecar-agent
    participant flagd-provider
    participant application

    flagd-provider-->proxy-sidecar-agent: Route traffic to backend sync service based on policy
    Note left of flagd-provider: In most cases the connection made to <br> `localhost` data plane port <br> e.g. localhost:9211

    flagd-provider-->flagd-sync.service: Get realtime flag update over gRPC stream
    Note left of proxy-sidecar-agent: Based on the host header i.e. authority <br> the tarrfic was then routed <br> backend service.
    loop
        flagd-provider->>flagd-provider: in-memory cache
    end
    application-->>flagd-provider: Check the state of a feature flag
    flagd-provider-->>application: Get the feature flag from in-memory cache <br/> run the evaluation logic and return final state

> ```

Reviewer 2: This diagram is a bit unclear to me due to the lack of directional arrows.

Something more like https://grpc.io/docs/guides/custom-name-resolution/#life-of-a-target-string makes sense to me.

Author: do you mean replace the diagram with the above one ?

Reviewer 2: This is good!

**Note:** There is small variation in supported alternative resolver e.g. java support `zooKeeper`

## Motivation

The main motivation is to support complex deployments with a generic custom name resolver using the interface
provided by gRPC core*.

**Note**: As of now only `java` and `golang` has the required interface to create custom resolver

## Detailed design

The idea is to

> * create custom name resolver for `java` and `golang` based on the example [provided](https://grpc.io/docs/guides/custom-name-resolution/#language-support)
> * and allow a new config option to pass the [target](https://grpc.io/docs/guides/custom-name-resolution/#life-of-a-target-string)
> string

Reviewer 2: ```suggestion
- allow a new config option to pass the [target](https://grpc.io/docs/guides/custom-name-resolution/#life-of-a-target-string) string
- reduce need to create/override existing implementations to simplify use of name-resolver
```

### Target String Pattern*

There is no restriction on naming but the string most comply with below standard

```text
scheme://authority/endpoint_name
```

#### Scheme Suggestions

We can choose our own scheme considering it's not used in gRPC core

| Scheme             | Comment | Status             |
|--------------------|---------|--------------------|
| **flagd://**       |         | Approved / Reject  |
| **grpc-remote://** |         | Approved / Reject  |

> | **tbd**            |         | Approved / Reject  |

Reviewer 2: Can you update this section to be less about your specific use case? Maybe just give an example scheme and what it might do in a hypothetical infrastructure.

Author: done replace the section with envoy sidecar example
